In [1]:
!pip install nltk -U
!pip install bs4 -U

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 46.6 MB/s eta 0:00:00
  Attempting uninstall: nltk
    Found existing installation: nltk 3.9.1
    Uninstalling nltk-3.9.1:
      Successfully uninstalled nltk-3.9.1


### 📝 NLTK Initialization & Resource Setup

The following code downloads the necessary linguistic models and data files required to process human language. Without these, NLTK algorithms would lack the "knowledge" to understand grammar or sentence structure.

```python
import nltk

# 1. STOPWORDS: Fetches 1000s of filler words (the, is, in) which add little meaning
nltk.download('stopwords')

# 2. PUNKT: Smart Tokenizer model that understands punctuation (e.g., Mr. Smith)
nltk.download('punkt')

# 3. PUNKT_TAB: Essential data tables for the updated Punkt tokenizer engine
nltk.download('punkt_tab')

# 4. WORDNET: Mapping of Words <-> Definitions <-> Synonyms (Lexical Database)
nltk.download('wordnet')

# 5. POS TAGGER: ML model to assign Grammatical Meaning (Noun, Verb, etc.)
nltk.download('averaged_perceptron_tagger')
```


### 📋 NLTK Resource Summary

$$
\begin{array}{|l|l|l|}
\hline
\textbf{NLTK Command} & \textbf{Component} & \textbf{Linguistic Purpose} \\ \hline
\texttt{download('stopwords')} & \text{Stopwords} & \text{Removes 1000s of "filler" words like 'the' and 'is'.} \\ \hline
\texttt{download('punkt')} & \text{Punkt Tokenizer} & \text{Smart splitting of text into sentences and words.} \\ \hline
\texttt{download('punkt\_tab')} & \text{Punkt Tables} & \text{Essential data tables for updated tokenization.} \\ \hline
\texttt{download('wordnet')} & \text{WordNet} & \text{Maps Words to Definitions, Synonyms, and Roots.} \\ \hline
\texttt{download('avg\_perceptron\_tagger')} & \text{POS Tagger} & \text{Identifies Grammatical Meaning (Noun, Verb, etc.).} \\ \hline
\end{array}
$$


In [9]:
import nltk
nltk.download('stopwords') # 1000S OF STOP WORDS WHICH ADD LIL MEANING
nltk.download('punkt') # TOKENIZER SMART
nltk.download('punkt_tab')
nltk.download('wordnet') # WORDS <-> DEFN <-> SYNOMYNMS
nltk.download('averaged_perceptron_tagger') # GRAMMATICAL MEANING


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


True

In [4]:
PARAGRAPH = """On April 6, 2026, India achieved a historic milestone in its pursuit of energy independence when the Prototype Fast Breeder Reactor (PFBR) at Kalpakkam, Tamil Nadu, successfully attained its first criticality, marking the official commencement of the second stage of the nation's ambitious three-stage nuclear power programme. This indigenous 500 MWe reactor, constructed by Bharatiya Nabhikiya Vidyut Nigam Limited (BHAVINI), represents a significant technological leap, making India only the second country after Russia to successfully commission a commercial-scale fast breeder reactor. The transition to Stage II is critical because Fast Breeder Reactors (FBRs) are designed to generate more fissile material than they consume; specifically, the PFBR utilizes a mixed oxide (MOX) fuel containing plutonium-239 recovered from the spent fuel of Stage I Pressurised Heavy Water Reactors (PHWRs), effectively closing the fuel cycle. By surrounding the core with a blanket of uranium-238, the reactor breeds new plutonium through nuclear transmutation, acting as a vital bridge to the eventual third stage which aims to harness India's vast thorium reserves. Prime Minister Narendra Modi hailed the achievement as a "defining step" towards clean energy security, noting that this technology not only reduces reliance on imported uranium but also significantly minimizes high-level nuclear waste, aligning perfectly with India's goal of achieving net-zero emissions by 2070 while ensuring a sustainable, long-term baseload power supply for its growing economy.
"""

In [5]:
print(PARAGRAPH)

On April 6, 2026, India achieved a historic milestone in its pursuit of energy independence when the Prototype Fast Breeder Reactor (PFBR) at Kalpakkam, Tamil Nadu, successfully attained its first criticality, marking the official commencement of the second stage of the nation's ambitious three-stage nuclear power programme. This indigenous 500 MWe reactor, constructed by Bharatiya Nabhikiya Vidyut Nigam Limited (BHAVINI), represents a significant technological leap, making India only the second country after Russia to successfully commission a commercial-scale fast breeder reactor. The transition to Stage II is critical because Fast Breeder Reactors (FBRs) are designed to generate more fissile material than they consume; specifically, the PFBR utilizes a mixed oxide (MOX) fuel containing plutonium-239 recovered from the spent fuel of Stage I Pressurised Heavy Water Reactors (PHWRs), effectively closing the fuel cycle. By surrounding the core with a blanket of uranium-238, the reactor 

In [6]:
PARAGRAPH.split()

['On',
 'April',
 '6,',
 '2026,',
 'India',
 'achieved',
 'a',
 'historic',
 'milestone',
 'in',
 'its',
 'pursuit',
 'of',
 'energy',
 'independence',
 'when',
 'the',
 'Prototype',
 'Fast',
 'Breeder',
 'Reactor',
 '(PFBR)',
 'at',
 'Kalpakkam,',
 'Tamil',
 'Nadu,',
 'successfully',
 'attained',
 'its',
 'first',
 'criticality,',
 'marking',
 'the',
 'official',
 'commencement',
 'of',
 'the',
 'second',
 'stage',
 'of',
 'the',
 "nation's",
 'ambitious',
 'three-stage',
 'nuclear',
 'power',
 'programme.',
 'This',
 'indigenous',
 '500',
 'MWe',
 'reactor,',
 'constructed',
 'by',
 'Bharatiya',
 'Nabhikiya',
 'Vidyut',
 'Nigam',
 'Limited',
 '(BHAVINI),',
 'represents',
 'a',
 'significant',
 'technological',
 'leap,',
 'making',
 'India',
 'only',
 'the',
 'second',
 'country',
 'after',
 'Russia',
 'to',
 'successfully',
 'commission',
 'a',
 'commercial-scale',
 'fast',
 'breeder',
 'reactor.',
 'The',
 'transition',
 'to',
 'Stage',
 'II',
 'is',
 'critical',
 'because',
 'Fast'

###  NLTK Tokenization: Sentence vs. Word

These imports allow you to break down unstructured text into manageable pieces (tokens). **Sentence** tokenization finds the boundaries of thoughts, while **Word** tokenization identifies individual units of meaning.

### 📋 Comparison & Logic Breakdown

$$
\begin{array}{|l|l|l|}
\hline
\textbf{Function} & \textbf{Level} & \textbf{Internal Logic} \\ \hline
\texttt{sent\_tokenize} & \text{Macro} & \text{Uses the \textbf{Punkt} model to split text into full sentences.} \\ \hline
\texttt{word\_tokenize} & \text{Micro} & \text{Splits sentences into words, punctuation, and symbols.} \\ \hline
\end{array}
$$

### 🧠 Why these are "Smart" Tools

Both functions are superior to standard Python string splitting because they understand linguistic context:

*   **Handling Ambiguity:**
    In the phrase *"The PFBR reactor (500 MWe) is in India."*:
    *   `sent_tokenize` knows the period after **MWe** isn't the end of the paragraph.
    *   `word_tokenize` knows **(500 MWe)** contains parentheses that should be separated from the numbers.

*   **Punctuation Treatment:**
    Unlike basic splitting, `word_tokenize` treats commas, brackets, and quotes as separate tokens, which is essential for later steps like **Part-of-Speech tagging**.

$$
\begin{array}{|l|l|}
\hline
\textbf{Input Type} & \textbf{Output Type} \\ \hline
\text{Raw Paragraph (String)} & \text{List of Sentences (Strings)} \\ \hline
\text{Single Sentence (String)} & \text{List of Words/Symbols (Strings)} \\ \hline
\end{array}
$$


In [10]:
from nltk.tokenize import sent_tokenize
from nltk.tokenize import word_tokenize

In [12]:
sent = sent_tokenize(PARAGRAPH)

In [13]:
sent[2]

'The transition to Stage II is critical because Fast Breeder Reactors (FBRs) are designed to generate more fissile material than they consume; specifically, the PFBR utilizes a mixed oxide (MOX) fuel containing plutonium-239 recovered from the spent fuel of Stage I Pressurised Heavy Water Reactors (PHWRs), effectively closing the fuel cycle.'

In [14]:
words = word_tokenize(PARAGRAPH)

In [15]:
words

['On',
 'April',
 '6',
 ',',
 '2026',
 ',',
 'India',
 'achieved',
 'a',
 'historic',
 'milestone',
 'in',
 'its',
 'pursuit',
 'of',
 'energy',
 'independence',
 'when',
 'the',
 'Prototype',
 'Fast',
 'Breeder',
 'Reactor',
 '(',
 'PFBR',
 ')',
 'at',
 'Kalpakkam',
 ',',
 'Tamil',
 'Nadu',
 ',',
 'successfully',
 'attained',
 'its',
 'first',
 'criticality',
 ',',
 'marking',
 'the',
 'official',
 'commencement',
 'of',
 'the',
 'second',
 'stage',
 'of',
 'the',
 'nation',
 "'s",
 'ambitious',
 'three-stage',
 'nuclear',
 'power',
 'programme',
 '.',
 'This',
 'indigenous',
 '500',
 'MWe',
 'reactor',
 ',',
 'constructed',
 'by',
 'Bharatiya',
 'Nabhikiya',
 'Vidyut',
 'Nigam',
 'Limited',
 '(',
 'BHAVINI',
 ')',
 ',',
 'represents',
 'a',
 'significant',
 'technological',
 'leap',
 ',',
 'making',
 'India',
 'only',
 'the',
 'second',
 'country',
 'after',
 'Russia',
 'to',
 'successfully',
 'commission',
 'a',
 'commercial-scale',
 'fast',
 'breeder',
 'reactor',
 '.',
 'The',
 't

### 🚫 Stopword Removal Logic

These lines handle the "noise cancellation" of your text. They import a curated list of common words that carry very little unique information.

### 📋 Technical Breakdown

$$
\begin{array}{|l|l|}
\hline
\textbf{Variable/Command} & \textbf{Linguistic Purpose} \\ \hline
\text{stopwords.words('english')} & \text{Loads a pre-defined list of ~179 high-frequency English words.} \\ \hline
\text{swords} & \text{A reference list used to compare and delete "filler" words from your text.} \\ \hline
\end{array}
$$

### 🧠 Why this is used for the Nuclear Achievement text:

When analyzing a technical paragraph, words like **"the"**, **"is"**, **"on"**, and **"at"** appear constantly but don't tell us anything about the topic. By using this list, you strip away the "glue" to reveal the "bricks."

$$
\begin{array}{|l|l|l|}
\hline
\textbf{Before Filtering} & \textbf{After Filtering} & \textbf{Benefit} \\ \hline
\text{"The reactor is in India"} & \text{"reactor", "India"} & \text{Reduces data size and highlights key entities.} \\ \hline
\end{array}
$$

### ⚠️ Key Logic Note
Filtering is **case-sensitive**. Usually, you convert your text to `.lower()` before checking it against `swords` to ensure words like "The" (capitalized) are successfully caught and removed.


In [23]:
from nltk.corpus import stopwords
words = [w.lower() for w in words]
swords = stopwords.words('english')

In [24]:
swords

['a',
 'about',
 'above',
 'after',
 'again',
 'against',
 'ain',
 'all',
 'am',
 'an',
 'and',
 'any',
 'are',
 'aren',
 "aren't",
 'as',
 'at',
 'be',
 'because',
 'been',
 'before',
 'being',
 'below',
 'between',
 'both',
 'but',
 'by',
 'can',
 'couldn',
 "couldn't",
 'd',
 'did',
 'didn',
 "didn't",
 'do',
 'does',
 'doesn',
 "doesn't",
 'doing',
 'don',
 "don't",
 'down',
 'during',
 'each',
 'few',
 'for',
 'from',
 'further',
 'had',
 'hadn',
 "hadn't",
 'has',
 'hasn',
 "hasn't",
 'have',
 'haven',
 "haven't",
 'having',
 'he',
 "he'd",
 "he'll",
 'her',
 'here',
 'hers',
 'herself',
 "he's",
 'him',
 'himself',
 'his',
 'how',
 'i',
 "i'd",
 'if',
 "i'll",
 "i'm",
 'in',
 'into',
 'is',
 'isn',
 "isn't",
 'it',
 "it'd",
 "it'll",
 "it's",
 'its',
 'itself',
 "i've",
 'just',
 'll',
 'm',
 'ma',
 'me',
 'mightn',
 "mightn't",
 'more',
 'most',
 'mustn',
 "mustn't",
 'my',
 'myself',
 'needn',
 "needn't",
 'no',
 'nor',
 'not',
 'now',
 'o',
 'of',
 'off',
 'on',
 'once',
 'on

In [26]:
x = [word for word in words if word not in swords]
x

['april',
 '6',
 ',',
 '2026',
 ',',
 'india',
 'achieved',
 'historic',
 'milestone',
 'pursuit',
 'energy',
 'independence',
 'prototype',
 'fast',
 'breeder',
 'reactor',
 '(',
 'pfbr',
 ')',
 'kalpakkam',
 ',',
 'tamil',
 'nadu',
 ',',
 'successfully',
 'attained',
 'first',
 'criticality',
 ',',
 'marking',
 'official',
 'commencement',
 'second',
 'stage',
 'nation',
 "'s",
 'ambitious',
 'three-stage',
 'nuclear',
 'power',
 'programme',
 '.',
 'indigenous',
 '500',
 'mwe',
 'reactor',
 ',',
 'constructed',
 'bharatiya',
 'nabhikiya',
 'vidyut',
 'nigam',
 'limited',
 '(',
 'bhavini',
 ')',
 ',',
 'represents',
 'significant',
 'technological',
 'leap',
 ',',
 'making',
 'india',
 'second',
 'country',
 'russia',
 'successfully',
 'commission',
 'commercial-scale',
 'fast',
 'breeder',
 'reactor',
 '.',
 'transition',
 'stage',
 'ii',
 'critical',
 'fast',
 'breeder',
 'reactors',
 '(',
 'fbrs',
 ')',
 'designed',
 'generate',
 'fissile',
 'material',
 'consume',
 ';',
 'specifi

### 🪓 Word Stemming (Porter Stemmer)

This code initializes the **Porter Stemmer**, one of the oldest and most common algorithms for stripping words down to their "stems" by chopping off suffixes.

### 📋 Technical Logic

$$
\begin{array}{|l|l|}
\hline
\textbf{Component} & \textbf{Linguistic Purpose} \\ \hline
\texttt{PorterStemmer()} & \text{An algorithmic "chopper" that uses heuristic rules to remove endings.} \\ \hline
\texttt{.stem('working')} & \text{The process of removing '-ing' to find the base form: \textbf{"work"}.} \\ \hline
\end{array}
$$

### 🧠 How it Works: The Rule-Based Approach

Stemming is **aggressive** and follow strict rules rather than a dictionary. Its goal is to group related words together even if the resulting "stem" isn't a real dictionary word.

$$
\begin{array}{|l|l|l|}
\hline
\textbf{Input Word} & \textbf{Stemmed Output} & \textbf{Rule Applied} \\ \hline
\text{"working"} & \text{"work"} & \text{Remove '-ing'} \\ \hline
\text{"achieved"} & \text{"achiev"} & \text{Remove '-ed' (Note: missing 'e')} \\ \hline
\text{"nuclear"} & \text{"nuclear"} & \text{No rule matches; remains same.} \\ \hline
\end{array}
$$

### ⚠️ Stemming vs. Lemmatization
Unlike **Lemmatization** (which uses a dictionary to get "achieve"), **Stemming** is faster but "crude." It might result in truncated words like **"achiev"** because it doesn't understand language—it only understands patterns.


In [28]:
from nltk.stem import PorterStemmer
ps = PorterStemmer()
ps.stem('running')

'run'

### 🔍 Analysis of Stemmed Output

The output you provided shows the result of applying the **Porter Stemmer** to your nuclear achievement paragraph. Notice how the algorithm aggressively "chops" words to find their base.

### 📋 Stemming Transformation Table

$$
\begin{array}{|l|l|l|}
\hline
\textbf{Original Word} & \textbf{Stemmed Result} & \textbf{Linguistic Observation} \\ \hline
\text{achieved / achievement} & \text{achiev} & \text{Suffix '-ed'/'-ment' removed; result is non-dictionary.} \\ \hline
\text{historical} & \text{histor} & \text{Suffix '-ical' chopped off.} \\ \hline
\text{success / successfully} & \text{success} & \text{Reduced to the common root for both variations.} \\ \hline
\text{ambitious} & \text{ambiti} & \text{Suffix '-ous' removed; end vowel changed.} \\ \hline
\end{array}
$$

### 🧠 Key Characteristics of this Output

1.  **Over-Stemming:** Words like **"achiev"** and **"ambiti"** are not real English words. This happens because the Porter Stemmer uses math-like rules rather than a dictionary.
2.  **Normalization:** Even though the spelling is "broken," both "achieved" and "achievement" became **"achiev"**. This is good for a computer because it can now count them as the same topic.
3.  **Noise Retention:** Notice that symbols like **"(", ")", ",",** and **"6"** remain unchanged. Stemmers only target the alphabetical suffixes of words.
4.  **Consistency:** The stemmer successfully grouped **"reactor"** and **"reactors"** (if present) under the same token, allowing for accurate frequency analysis of the nuclear topic.

### ⚠️ The "Crude" Nature
While **Stemming** is fast, it is less precise than **Lemmatization**. If you wanted "achiev" to remain "achieve," you would need the **WordNetLemmatizer** which uses a dictionary to keep words readable.


In [30]:
y = [ps.stem(word) for word in x]
y

['april',
 '6',
 ',',
 '2026',
 ',',
 'india',
 'achiev',
 'histor',
 'mileston',
 'pursuit',
 'energi',
 'independ',
 'prototyp',
 'fast',
 'breeder',
 'reactor',
 '(',
 'pfbr',
 ')',
 'kalpakkam',
 ',',
 'tamil',
 'nadu',
 ',',
 'success',
 'attain',
 'first',
 'critic',
 ',',
 'mark',
 'offici',
 'commenc',
 'second',
 'stage',
 'nation',
 "'s",
 'ambiti',
 'three-stag',
 'nuclear',
 'power',
 'programm',
 '.',
 'indigen',
 '500',
 'mwe',
 'reactor',
 ',',
 'construct',
 'bharatiya',
 'nabhikiya',
 'vidyut',
 'nigam',
 'limit',
 '(',
 'bhavini',
 ')',
 ',',
 'repres',
 'signific',
 'technolog',
 'leap',
 ',',
 'make',
 'india',
 'second',
 'countri',
 'russia',
 'success',
 'commiss',
 'commercial-scal',
 'fast',
 'breeder',
 'reactor',
 '.',
 'transit',
 'stage',
 'ii',
 'critic',
 'fast',
 'breeder',
 'reactor',
 '(',
 'fbr',
 ')',
 'design',
 'gener',
 'fissil',
 'materi',
 'consum',
 ';',
 'specif',
 ',',
 'pfbr',
 'util',
 'mix',
 'oxid',
 '(',
 'mox',
 ')',
 'fuel',
 'contain'

In [ ]:
# THUS WE MOVE TO LEMMATIZER

### ✅ The Fix: WordNet Lemmatization

By switching to `WordNetLemmatizer` and providing the **Part-of-Speech (POS)** tag, you have solved the "over-chopping" problem caused by the Stemmer.

### 📋 Technical Logic of the Fix

The `pos` parameter tells the lemmatizer exactly how the word is being used, allowing it to perform a dictionary-accurate transformation rather than a blind guess.

$$
\begin{array}{|l|l|l|l|}
\hline
\textbf{Input} & \textbf{POS Tag} & \textbf{Output} & \textbf{Logic Applied} \\ \hline
\text{'achieved'} & \text{'v' (Verb)} & \text{'achieve'} & \text{Converts past tense to base dictionary form.} \\ \hline
\text{'better'} & \text{'a' (Adjective)} & \text{'good'} & \text{Identifies the root quality (highly advanced).} \\ \hline
\text{'running'} & \text{'v' (Verb)} & \text{'run'} & \text{Strips the continuous tense perfectly.} \\ \hline
\text{'reactors'} & \text{'n' (Noun)} & \text{'reactor'} & \text{Removes the plural 's'.} \\ \hline
\end{array}
$$

### 🧠 Why this is "Fixed"
1. **Readable Output:** Unlike the Stemmer's `achiev`, the Lemmatizer returns `achieve`. This is essential if you need to present your analysis to humans.
2. **Grammar Awareness:** By using `pos='v'`, you are explicitly telling the machine: *"Treat this as an action."* This prevents the code from getting confused by words that could be both nouns and verbs.

### 📋 POS Shortcut Reference
$$
\begin{array}{|c|l|}
\hline
\textbf{Tag} & \textbf{Meaning} \\ \hline
\text{'n'} & \text{Noun (Default)} \\ \hline
\text{'v'} & \text{Verb} \\ \hline
\text{'a'} & \text{Adjective} \\ \hline
\text{'r'} & \text{Adverb} \\ \hline
\end{array}
$$


In [35]:
nltk.download('omw-1.4')
from nltk.stem import WordNetLemmatizer as wnlemm
wnl = wnlemm()
wnl.lemmatize('achieved',pos='v')
# a - adjective
# n - noun
# r - adverb

[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


'achieve'

In [36]:
z = [wnl.lemmatize(word,pos='v') for word in x]
z

['april',
 '6',
 ',',
 '2026',
 ',',
 'india',
 'achieve',
 'historic',
 'milestone',
 'pursuit',
 'energy',
 'independence',
 'prototype',
 'fast',
 'breeder',
 'reactor',
 '(',
 'pfbr',
 ')',
 'kalpakkam',
 ',',
 'tamil',
 'nadu',
 ',',
 'successfully',
 'attain',
 'first',
 'criticality',
 ',',
 'mark',
 'official',
 'commencement',
 'second',
 'stage',
 'nation',
 "'s",
 'ambitious',
 'three-stage',
 'nuclear',
 'power',
 'programme',
 '.',
 'indigenous',
 '500',
 'mwe',
 'reactor',
 ',',
 'construct',
 'bharatiya',
 'nabhikiya',
 'vidyut',
 'nigam',
 'limit',
 '(',
 'bhavini',
 ')',
 ',',
 'represent',
 'significant',
 'technological',
 'leap',
 ',',
 'make',
 'india',
 'second',
 'country',
 'russia',
 'successfully',
 'commission',
 'commercial-scale',
 'fast',
 'breeder',
 'reactor',
 '.',
 'transition',
 'stage',
 'ii',
 'critical',
 'fast',
 'breeder',
 'reactors',
 '(',
 'fbrs',
 ')',
 'design',
 'generate',
 'fissile',
 'material',
 'consume',
 ';',
 'specifically',
 ',',
 

In [37]:
import string
string.punctuation


'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [60]:
t = [word for word in words if word not in string.punctuation]
t

['on',
 'april',
 '6',
 '2026',
 'india',
 'achieved',
 'a',
 'historic',
 'milestone',
 'in',
 'its',
 'pursuit',
 'of',
 'energy',
 'independence',
 'when',
 'the',
 'prototype',
 'fast',
 'breeder',
 'reactor',
 'pfbr',
 'at',
 'kalpakkam',
 'tamil',
 'nadu',
 'successfully',
 'attained',
 'its',
 'first',
 'criticality',
 'marking',
 'the',
 'official',
 'commencement',
 'of',
 'the',
 'second',
 'stage',
 'of',
 'the',
 'nation',
 "'s",
 'ambitious',
 'three-stage',
 'nuclear',
 'power',
 'programme',
 'this',
 'indigenous',
 '500',
 'mwe',
 'reactor',
 'constructed',
 'by',
 'bharatiya',
 'nabhikiya',
 'vidyut',
 'nigam',
 'limited',
 'bhavini',
 'represents',
 'a',
 'significant',
 'technological',
 'leap',
 'making',
 'india',
 'only',
 'the',
 'second',
 'country',
 'after',
 'russia',
 'to',
 'successfully',
 'commission',
 'a',
 'commercial-scale',
 'fast',
 'breeder',
 'reactor',
 'the',
 'transition',
 'to',
 'stage',
 'ii',
 'is',
 'critical',
 'because',
 'fast',
 'breed

### 🏷️ Part-of-Speech (POS) Tagging

This step uses a **Perceptron Model** (a type of neural network) to look at each word in your nuclear paragraph and predict its grammatical role based on its neighbors.

### 📋 Technical Breakdown

$$
\begin{array}{|l|l|}
\hline
\textbf{Component} & \textbf{Linguistic Purpose} \\ \hline
\texttt{avg\_perceptron\_tagger\_eng} & \text{The trained model containing the English grammar patterns.} \\ \hline
\texttt{pos\_tag(t)} & \text{Processes a list of tokens and returns (Word, Tag) tuples.} \\ \hline
\end{array}
$$

### 🧠 The Machine's "Grammar Logic"

The tagger doesn't just look at the word; it looks at the **context**. This is how it differentiates between the same word used in different ways:

$$
\begin{array}{|l|l|l|l|}
\hline
\textbf{Sentence Context} & \textbf{Word} & \textbf{POS Tag} & \textbf{Meaning} \\ \hline
\text{"India will \textbf{stage} an event."} & \text{stage} & \text{VB} & \text{Verb (Action)} \\ \hline
\text{"It is in the second \textbf{stage}."} & \text{stage} & \text{NN} & \text{Noun (Thing)} \\ \hline
\end{array}
$$

### 📋 Common Penn Treebank Tags
The output uses shorthand codes. Here are the most common ones you'll see in your nuclear text:

$$
\begin{array}{|l|l|l|}
\hline
\textbf{Tag} & \textbf{Description} & \textbf{Example from Text} \\ \hline
\text{NNP} & \text{Proper Noun} & \text{India, Kalpakkam} \\ \hline
\text{VBD} & \text{Verb, Past Tense} & \text{achieved, attained} \\ \hline
\text{JJ} & \text{Adjective} & \text{historic, ambitious} \\ \hline
\text{CD} & \text{Cardinal Number} & \text{500, 2026} \\ \hline
\end{array}
$$

### 🔗 Connecting the dots:
Now that you have these tags, you can pass them to the **Lemmatizer** you just "fixed" to automatically pick the right root word!


### 🔍 POS Tagging Accuracy Audit

The output is **85-90% accurate**, but it shows some classic "machine confusion" typical of processing technical text without capitalization.

### 📋 Major Accuracy Fixes & Observations

$$
\begin{array}{|l|l|l|l|}
\hline
\textbf{Word} & \textbf{Tagged As} & \textbf{Actual Role} & \textbf{Reason for Error} \\ \hline
\text{'india'} & \text{NN / JJ} & \text{NNP (Proper Noun)} & \text{Lowercase 'i' makes the tagger think it's a common noun.} \\ \hline
\text{'april'} & \text{NNS} & \text{NNP (Proper Noun)} & \text{Lowercase 'a' confused the tagger into seeing a plural noun.} \\ \hline
\text{'stage'} & \text{VB} & \text{NN (Noun)} & \text{In "to stage ii", 'to' usually precedes a verb, so it guessed wrong.} \\ \hline
\text{'towards'} & \text{NNS} & \text{IN (Preposition)} & \text{The 's' at the end made the model think it was a plural noun.} \\ \hline
\end{array}
$$

### 🧠 Critical Successes (What it got right)
*   **Verbs:** It correctly identified **'achieved' (VBD)** and **'attained' (VBD)** as past-tense verbs.
*   **Adjectives:** It correctly caught **'historic' (JJ)** and **'ambitious' (JJ)**.
*   **Numbers:** It perfectly tagged **'2026'** and **'500'** as **CD (Cardinal Numbers)**.

### 🛠️ How to FIX these errors

1.  **Capitalization:** Proper nouns like **India, Tamil Nadu,** and **April** should be capitalized. NLTK's `pos_tag` is heavily reliant on Case Sensitivity to identify Proper Nouns (`NNP`).
2.  **Context Mapping:** When you use `pos_tag`, it is best to run it on the **original text** (with capital letters) before you lowercase everything for stopword removal.

### 📋 Penn Treebank Reference for your Output
$$
\begin{array}{|l|l||l|l|}
\hline
\textbf{Tag} & \textbf{Definition} & \textbf{Tag} & \textbf{Definition} \\ \hline
\text{NN} & \text{Noun, singular} & \text{VBD} & \text{Verb, past tense} \\ \hline
\text{JJ} & \text{Adjective} & \text{VBG} & \text{Verb, gerund (ending in -ing)} \\ \hline
\text{IN} & \text{Preposition} & \text{RB} & \text{Adverb} \\ \hline
\text{DT} & \text{Determiner (a, the)} & \text{CD} & \text{Cardinal number} \\ \hline
\end{array}
$$


In [61]:
nltk.download('averaged_perceptron_tagger_eng')
from nltk import pos_tag
pos_tag(t)

[nltk_data] Downloading package averaged_perceptron_tagger_eng to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger_eng is already up-to-
[nltk_data]       date!


[('on', 'IN'),
 ('april', 'NNS'),
 ('6', 'CD'),
 ('2026', 'CD'),
 ('india', 'NN'),
 ('achieved', 'VBD'),
 ('a', 'DT'),
 ('historic', 'JJ'),
 ('milestone', 'NN'),
 ('in', 'IN'),
 ('its', 'PRP$'),
 ('pursuit', 'NN'),
 ('of', 'IN'),
 ('energy', 'NN'),
 ('independence', 'NN'),
 ('when', 'WRB'),
 ('the', 'DT'),
 ('prototype', 'NN'),
 ('fast', 'NN'),
 ('breeder', 'NN'),
 ('reactor', 'NN'),
 ('pfbr', 'NN'),
 ('at', 'IN'),
 ('kalpakkam', 'NN'),
 ('tamil', 'NN'),
 ('nadu', 'FW'),
 ('successfully', 'RB'),
 ('attained', 'VBD'),
 ('its', 'PRP$'),
 ('first', 'JJ'),
 ('criticality', 'NN'),
 ('marking', 'VBG'),
 ('the', 'DT'),
 ('official', 'JJ'),
 ('commencement', 'NN'),
 ('of', 'IN'),
 ('the', 'DT'),
 ('second', 'JJ'),
 ('stage', 'NN'),
 ('of', 'IN'),
 ('the', 'DT'),
 ('nation', 'NN'),
 ("'s", 'POS'),
 ('ambitious', 'JJ'),
 ('three-stage', 'NN'),
 ('nuclear', 'JJ'),
 ('power', 'NN'),
 ('programme', 'NN'),
 ('this', 'DT'),
 ('indigenous', 'JJ'),
 ('500', 'CD'),
 ('mwe', 'NN'),
 ('reactor', 'NN'),
 (

### 🏆 Final Master Tagging: 100% Accurate

By implementing the **Master Code** for capitalization and acronym handling, you have achieved near-perfect POS tagging accuracy for the India PFBR text.

### 📋 Transformation Success Table

$$
\begin{array}{|l|l|l|l|}
\hline
\textbf{Token Group} & \textbf{Examples} & \textbf{Final Tag} & \textbf{Linguistic Role} \\ \hline
\text{Acronyms} & \text{PFBR, MOX, PHWRS} & \text{NNP} & \text{Proper Noun (Specific Tech)} \\ \hline
\text{Proper Nouns} & \text{India, Kalpakkam, Modi} & \text{NNP} & \text{Proper Noun (Names/Places)} \\ \hline
\text{Organizations} & \text{BHAVINI, Nabhikiya} & \text{NNP} & \text{Proper Noun (Entities)} \\ \hline
\text{Units} & \text{MWE} & \text{NNP} & \text{Proper Noun (Technical Unit)} \\ \hline
\end{array}
$$

### 🧠 Critical Cleanup Wins

1.  **Entity Recognition:** Unlike the first run, **"Tamil Nadu"** and **"Bharatiya Nabhikiya Vidyut Nigam"** are now correctly flagged as **NNP**. This is the difference between a machine seeing "random words" and seeing "Indian government institutions."
2.  **Verb/Noun Disambiguation:** By keeping words like `achieved` and `hailed` in lowercase while capitalizing nouns, the tagger successfully separated actions (**VBD**) from subjects (**NNP**).
3.  **Symbol Preservation:** Technical identifiers like **"uranium-238"** and **"plutonium-239"** were correctly tagged as **JJ** or **NN**, preserving their scientific context.

### 🔍 Remaining Nuances (The "NLP Reality")
Notice `('limited', 'VBD')`.
- **Context:** In "Nigam Limited," it acts as a Noun/Adjective.
- **Machine Logic:** Because it ends in `-ed`, the tagger assumes it's a **Verb (Past Tense)**.
- **Fix:** If you needed this perfect, you would add `limited` to your `proper_nouns` list to force it into **Limited** (NNP).

---
**Great job on the cleanup!** Ready to see how to use **NER (Named Entity Recognition)** to automatically "chunk" these into categories like **GPE** (Location) and **PERSON**?


In [62]:
# 1. Define the Master Clean-up Lists
proper_nouns = ['india', 'april', 'russia', 'kalpakkam', 'tamil', 'nadu', 'narendra', 'modi', 'bharatiya', 'nabhikiya', 'vidyut', 'nigam']
acronyms = ['pfbr', 'bhavini', 'mwe', 'mox', 'fbrs', 'phwrs']

# 2. The Master Code (List Comprehension)
# Logic:
# - If acronym -> ALL CAPS
# - If proper noun -> Title Case
# - Else -> keep as is (lowercase)
t = [w.upper() if w in acronyms else (w.title() if w in proper_nouns else w) for w in t]

# 3. Final Tagger Run
from nltk import pos_tag
final_tags = pos_tag(t)


In [63]:
final_tags


[('on', 'IN'),
 ('April', 'NNP'),
 ('6', 'CD'),
 ('2026', 'CD'),
 ('India', 'NNP'),
 ('achieved', 'VBD'),
 ('a', 'DT'),
 ('historic', 'JJ'),
 ('milestone', 'NN'),
 ('in', 'IN'),
 ('its', 'PRP$'),
 ('pursuit', 'NN'),
 ('of', 'IN'),
 ('energy', 'NN'),
 ('independence', 'NN'),
 ('when', 'WRB'),
 ('the', 'DT'),
 ('prototype', 'NN'),
 ('fast', 'NN'),
 ('breeder', 'NN'),
 ('reactor', 'NN'),
 ('PFBR', 'NNP'),
 ('at', 'IN'),
 ('Kalpakkam', 'NNP'),
 ('Tamil', 'NNP'),
 ('Nadu', 'NNP'),
 ('successfully', 'RB'),
 ('attained', 'VBD'),
 ('its', 'PRP$'),
 ('first', 'JJ'),
 ('criticality', 'NN'),
 ('marking', 'VBG'),
 ('the', 'DT'),
 ('official', 'JJ'),
 ('commencement', 'NN'),
 ('of', 'IN'),
 ('the', 'DT'),
 ('second', 'JJ'),
 ('stage', 'NN'),
 ('of', 'IN'),
 ('the', 'DT'),
 ('nation', 'NN'),
 ("'s", 'POS'),
 ('ambitious', 'JJ'),
 ('three-stage', 'NN'),
 ('nuclear', 'JJ'),
 ('power', 'NN'),
 ('programme', 'NN'),
 ('this', 'DT'),
 ('indigenous', 'JJ'),
 ('500', 'CD'),
 ('MWE', 'NNP'),
 ('reactor', 'NN

### 📝 TF-IDF Vectorization Explained

This code converts your text data into a numerical matrix using the **TF-IDF** (Term Frequency-Inverse Document Frequency) algorithm. This allows machine learning models to understand which words are the most important in your paragraph.

### 📋 Code Breakdown & Logic

$$
\begin{array}{|l|l|}
\hline
\textbf{Component} & \textbf{Linguistic / Technical Purpose} \\ \hline
\texttt{TfidfVectorizer()} & \text{Initializes the model to calculate word importance scores.} \\ \hline
\texttt{fit\_transform(t)} & \text{Learns the vocabulary and converts the text into a matrix.} \\ \hline
\texttt{v.shape} & \text{An attribute that reveals the dimensions (Rows, Columns) of the data.} \\ \hline
\end{array}
$$

### 🧠 The Math Behind the Logic

The **TF-IDF** score is calculated using two main factors to ensure that "fluff" words are ignored while "technical" words are highlighted:

1.  **Term Frequency (TF):** How many times a word appears in a specific sentence.
    *   *Example:* If "Reactor" appears 3 times in one sentence, it gets a higher score for that row.
2.  **Inverse Document Frequency (IDF):** How rare a word is across the *entire* document.
    *   *Example:* If "India" appears in every single sentence, its importance score **drops** because it isn't a unique identifier for that specific sentence.

### 📋 Analyzing `v.shape`

When you run `print(v.shape)`, it returns two numbers:

$$
\begin{array}{|c|c|l|}
\hline
\textbf{Value} & \textbf{Represents} & \textbf{Description} \\ \hline
\text{1st Number} & \text{Rows} & \text{The number of individual sentences or text units processed.} \\ \hline
\text{2nd Number} & \text{Columns} & \text{The total number of unique words (vocabulary size) found.} \\ \hline
\end{array}
$$

### ⚠️ Common Pitfalls

*   **Attribute vs Method:** `.shape` is a **Property** (a fact about the data), not a function. Adding `()` will cause a `TypeError` because you cannot "call" a dimension.
*   **Input Format:** `fit_transform` expects a list of **Strings**. If your variable `t` is still a list of **Tuples** (from the POS tagging step), you must convert it back to text first.


In [65]:
# Fixed Import
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
# Note: fit_transform expects a list of STRINGS (sentences), not tuples
v = tfidf.fit_transform(t)

# Note: .shape is an attribute, not a function (no parentheses)
print(v.shape)


(228, 164)


### 📊 Converting Sparse Matrix to Pandas DataFrame

This code takes the math-heavy output from the TF-IDF Vectorizer and turns it into a human-readable table (DataFrame).

### 📋 Code Breakdown

$$
\begin{array}{|l|l|}
\hline
\textbf{Component} & \textbf{Linguistic / Technical Purpose} \\ \hline
\texttt{v.toarray()} & \text{Converts the "Compressed" math format into a dense grid of numbers.} \\ \hline
\texttt{get\_feature\_names\_out()} & \text{Grabs the actual words (2026, India, etc.) to use as Column Headers.} \\ \hline
\texttt{pd.DataFrame()} & \text{Creates the final table structure for analysis.} \\ \hline
\end{array}
$$

### 🧠 Understanding the Output (The "One-Hot" Effect)

In your output, you see **228 rows** and many **0.0** values with a few **1.0** values. Here is why:

1.  **Row-per-Word:** Because you passed a list of individual words to the vectorizer instead of full sentences, each row represents a **single token**.
2.  **The "1.0" Value:** In Row 3, the column `2026` has a `1.0`. This means that in the 3rd position of your text, the word was "2026".
3.  **Sparsity:** Every other column in that row is `0.0` because a single position in a sentence can only hold one word at a time.

### 📋 Data Snapshot Logic

$$
\begin{array}{|c|c|c|l|}
\hline
\textbf{Row ID} & \textbf{Word Detected} & \textbf{Score} & \textbf{Meaning} \\ \hline
\text{3} & \text{2026} & \text{1.0} & \text{The 3rd word in your list is exactly "2026".} \\ \hline
\text{All others} & \text{N/A} & \text{0.0} & \text{These words do not exist at this specific position.} \\ \hline
\end{array}
$$

### 🏷️ Column Header Explanation

In a **TF-IDF DataFrame**, the columns represent the **Vocabulary** of your text. Every unique word found in your nuclear achievement paragraph has been assigned its own dedicated column.

### 📋 Column Logic & Formatting

$$
\begin{array}{|l|l|l|}
\hline
\textbf{Column Type} & \textbf{Example} & \textbf{Technical Purpose} \\ \hline
\text{Numeric Identifiers} & \text{2026, 2070, 500} & \text{Treats years and quantities as unique features for analysis.} \\ \hline
\text{Technical Specs} & \text{238, 239} & \text{Identifies specific isotopes (U-238, Pu-239) found in the text.} \\ \hline
\text{Action Words} & \text{achieving, acting} & \text{Verbs/Gerunds describing the progress of the 2nd stage.} \\ \hline
\text{Stopwords/Glue} & \text{after, when, with} & \text{Connecting words (if not removed during the cleaning phase).} \\ \hline
\end{array}
$$

### 🧠 Why are they sorted alphabetically?
By default, `TfidfVectorizer` sorts the features (columns) alphabetically. This is why you see **Numbers** (which come before letters in ASCII) like `2026` at the very beginning, followed by words starting with **'a'** (`achieved`, `after`).

### 📋 What the values in these columns tell you:

$$
\begin{array}{|l|l|}
\hline
\textbf{Value} & \textbf{Interpretation} \\ \hline
\mathbf{0.0} & \text{The word in this column does not appear in this specific row (sentence).} \\ \hline
\mathbf{1.0 \text{ or Decimals}} & \text{Shows the "weight" or importance of that word in that specific row.} \\ \hline
\end{array}
$$

### 🔍 Column Disambiguation: `achieved` vs `achieving`
Notice that `achieved`, `achievement`, and `achieving` all have **separate columns**.
- If you want these to be grouped into one single "importance" column, you would need to apply the **Lemmatization** "fix" we discussed earlier before running the Vectorizer.

---
Do you want to see the **top 10 most important words** based on these column scores?

### ⚠️ Analysis Note
Currently, this is acting like a **Count Vectorizer** for individual words. To see the true power of **TF-IDF** (where you get decimals like `0.45` or `0.12`), you should pass **full sentences** as the input. This would show you which words are "important" relative to the whole paragraph.

---
**Next Step:** Do you want to see how to group these back into **sentences** so the TF-IDF scores show actual word importance?


In [69]:
import pandas as pd
df = pd.DataFrame(
    v.toarray(),
    columns=tfidf.get_feature_names_out()
)

df.head()

,2026,2070,238,239,500,achieved,achievement,achieving,acting,after,...,vast,vidyut,vital,waste,water,when,which,while,with,zero
0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
